endpoints/openrtb2/auction_test.go

```diff
@@ -36,19 +37,244 @@ import (
+	}
+	if len(tc.AdapterList) > 0 {
+		tc.adaptersConfig = make(map[string]config.Adapter, len(tc.AdapterList))
+		for _, adapterName := range tc.AdapterList {
+			tc.adaptersConfig[adapterName] = config.Adapter{Disabled: true}
+		}
+	}
+	return tc
+}
+
+// This function does not expects errors when unmarshaling both the actual and the expected bid response
+// Once unmarshalled, bidResponse objects can't simply be compared with an `assert.Equalf()` call
+// because tests fail if the elements inside the `bidResponse.SeatBid` and `bidResponse.SeatBod.Bid`
+// arrays, if any, are not listed in the exact same order in the actual version and in the expected version.
+func assertBidResponseEqual(t *testing.T, testFile string, actualJson []byte, expectedJson []byte) {
+	var actualBidResponse openrtb.BidResponse
+	var expectedBidResponse openrtb.BidResponse
+
+	if err := json.Unmarshal(actualJson, &actualBidResponse); err != nil {
+		t.Fatalf("Could not unmarshal actual bidResponse from auction.\n Test file: %s\n Error:%s\n", testFile, err)
+	}
+
+	if err := json.Unmarshal(expectedJson, &expectedBidResponse); err != nil {
+		t.Fatalf("Could not unmarshal expected bidResponse taken from test file.\n Test file: %s\n Error:%s\n", testFile, err)
+	}
+
+	assert.Equalf(t, actualBidResponse.ID, expectedBidResponse.ID, "BidResponse.ID doesn't match expected. Test: %s\n", testFile)
+	assert.Len(t, actualBidResponse.SeatBid, len(expectedBidResponse.SeatBid), "BidResponse.SeatBid array doesn't match expected. Test: %s\n", testFile)
+
+	//Given that bidResponses have the same lenght, compare them
+	for i := 0; i < len(expectedBidResponse.SeatBid); i++ {
+		//Map SeatBid array elements to get an accurate comparison
+		var actualSeatBidsMap map[string]openrtb.SeatBid = make(map[string]openrtb.SeatBid, 0)
+		for _, seat := range actualBidResponse.SeatBid {
+			actualSeatBidsMap[seat.Seat] = seat
+		}
+
+		var expectedSeatBidsMap map[string]openrtb.SeatBid = make(map[string]openrtb.SeatBid, 0)
+		for _, seat := range expectedBidResponse.SeatBid {
+			expectedSeatBidsMap[seat.Seat] = seat
+		}
+
+		for k, expectedSeatBid := range expectedSeatBidsMap {
+			assert.Len(t, actualSeatBidsMap[k].Bid, len(expectedSeatBid.Bid), "BidResponse.SeatBid[].Bid array doesn't match expected. Test: %s\n", testFile)
+
+			//Map bids to get actualBidResponse.SeatBid[i].Bid elements sorted so order doesn't matter when comparing
+			var actualBidsMap map[string]openrtb.Bid = make(map[string]openrtb.Bid, 0)
+			for _, bid := range actualSeatBidsMap[k].Bid {
+				actualBidsMap[bid.ID] = bid
+			}
+
+			var expectedBidsMap map[string]openrtb.Bid = make(map[string]openrtb.Bid, 0)
+			for _, bid := range expectedSeatBid.Bid {
+				expectedBidsMap[bid.ID] = bid
+			}
+
+			//Compare maps order in which bids were listed in the BidResponse.SeatBid[%d].Bid is irrelevant
+			assert.Equalf(t, actualBidsMap, expectedBidsMap, "BidResponse.SeatBid[].Bid doesn't match expected. Test: %s\n", testFile)
+		}
```

Comment:
Can you please explain a bit what's going on in this compare? There is some magic in Go, where comparing maps with assert.Equals does an order-irrelevant deep compare.

Reply:
The main reason I had to create this function is accurately compare even if `[]SeatBid` array and the `[]Bid` arrays within the `SeatBid` elements come in different order. Function `assert.Equals(t, expBidResponse, actualBidResponse)` fails if the `SeatBid` elements are listed in different order even if they are identical. Simply comparing `expectedSeatBidsMap` vs `actualSeatBidsMap` with `assert.Equals(...)` could also lead to false negatives if the elements in their `[]Bid` arrays come in different order too. To clarify further I added the `TestBidRequestAssert` test case. Please let me know if you believe there's a simpler way to compare accurately independent of both the `[]SeatBid` array element order and their respective `[]Bid` array element order or if this is even desired.

I refactored this test case a bit and added more comments that probably make it a bit more readable.

Reply:
Ah, ok. The normal compares fail due to an array embedded within the map. Fun. I can't think of a better way to do this.